Rework parsing: More strict and (hopefully) cleaner

Lingepumpe opened the pull request:
Reworked every parser file, with the exception of extensions.rs.

Changes and cleanup were mostly:
- Every parser is now responsible for consuming his own starting and closing tags (previously some
  parsers would be okay not having any starting tag, and would be used like this). To be able to achieve
  this any parser that calls further consume functions must use peek() to leave starting tags in the reader
  for the subparser to handle.
- Many parsers were accepting any amount of their starting tag, and also consuming any closing tag as
  valid, ignoring the name. Now exactly one opening tag and matching closing tag is required.
- Parsers that were already using peek() were doing so via a level of indirection through some form of
  "ParseEvent" enum, I think I have found a cleaner solution using .clone() on the event which allows us
  to remove this indirection and parse XmlEvents more directly even with peek().
- Added some more custom errors, and generally use custom errors more

brendanashworth commented on src/types.rs:
```diff
@@ -171,12 +171,26 @@ impl ToGeo<f64> for TrackSegment {
 
+// A Version of geo::Point that has the Default trait implemented
+#[derive(Clone, Debug)]
+struct OurPoint {
```
Would using a [one-field tuple](https://doc.rust-lang.org/1.6.0/book/structs.html#tuple-structs) here be preferable/cleaner? Also, please rename it to `GpxPoint` and add a comment explaining why we added default in the first place.

```rust
struct GpxPoint(Point);

impl Default for GpxPoint {
    fn default() -> GpxPoint {
        GpxPoint(Point::new(0., 0.))
    }
}
```

brendanashworth commented on src/parser/mod.rs:
```diff
@@ -45,3 +54,33 @@ impl<R: Read> Context<R> {
+            })) => {
+                ensure!(
+                    name.local_name == local_name,
+                    ErrorKind::InvalidChildElement(name.local_name, local_name)
```
Technically speaking, this would give an error of "invalid child element 'wpt' in 'wpt'", right? Because we don't keep track of the parent element?

Or would this even happen because we check the name beforehand, so instead we should be dying here or something?

Lingepumpe replied:
I dont think it should be wpt in wpt, note the first is name.local_name, which is the actual starting tag we found, and the other is "local_name", which is the expected local_name (the starting tag we are looking for). So I think what would happen, when looking for the wpt starting tag and getting trk starting tag instead is "invalid child element trk in wpt".

I agree this is stretching the InvalidChildElement error a bit, we could add a InvalidStartingElement error that would read something like "invalid starting element trk found, expected wpt". Note that the way the parser currently functions it shouldnt be possible to produce this error except in gpx - because all other parsers are sub-parser, which would only be called if the parent parser already found their starting tag via peek().

brendanashworth replied:
I see, that makes sense. It shouldn't need changing then, at least for now. Thanks for the clarity.

brendanashworth commented on src/parser/mod.rs:
```diff
@@ -45,3 +54,33 @@ impl<R: Read> Context<R> {
+            Some(Ok(XmlEvent::Characters(chars))) => {
+                bail!(ErrorKind::InvalidChildElement(chars, local_name));
+            }
+            Some(_) => {} //ignore other elements
```
What kind of event would we be skipping here, shouldn't it always be a `StartElement` at this point?

Lingepumpe replied:
If you look at https://netvl.github.io/xml-rs/xml/reader/enum.XmlEvent.html you can see there are a lot of events possible. My "set parser config when spanwing the XmlReader" commit that I added below will at least turn XmlEvent::Whitespace and XmlEvent::CData to become XmlEvent::Character. XmlEvent::Comment is disabled by default.

This leaves us with these possible events for the Some(_) line: StartDocument, EndDocument and ProcessingInstruction. Especially StartDocument will always occur at the beginning of any xml document (even if there is no explicit StartDocument tag), so it is always emitted in our test functions. EndDocument we could handle and bail! on it, but we can also just ignore it and get "None" in the next loop iteration. Finally ProcessingInstruction sounds weird, I am not sure how they work or if they exist for gpx, so for now I thought it safe to ignore.

brendanashworth replied:
Interesting, thanks for clearing that up!

brendanashworth commented on src/parser/gpx.rs:
```diff
@@ -260,6 +190,7 @@ mod tests {
             GpxVersion::Unknown
         );
 
+        println!("########################### {:?}", gpx);
```
Should we keep this?

Lingepumpe commented:
Note sure what happened with the Travis build on beta. Did I cause this?

Lingepumpe commented:
I believe I addressed all review concerns, or is there something else you would like me to fix? Or will some other person review as well? Just checking the ball is not in my court ;)

brendanashworth commented:
@Lingepumpe sorry about the delay, I've been unable to compile your changes due to some issues with (presumably) my Rust setup, and I've also been very busy. I should have time to do a final review and merge this within the next two days, thanks for being patient. This looks really nice though.

Lingepumpe commented:
It seems cargo fmt is now more restrictive with regards to "use": They are ordered in a certain way, and I also got more duplicate warnings. Updated the pull request to include those fixes.

frewsxcv commented:
@Lingepumpe thanks for fixing the cargo fmt issue

@brendanashworth still planning to look at this? otherwise, i'm thinking about merging this tomorrow

frewsxcv commented:
bors r+
